nibabel/streamlines/compact_list.py

```diff
@@ -0,0 +1,232 @@
+            self._offsets = np.array([0])
+            self._lengths = np.array([len(element)])
+            return
+
+        if element.shape[1:] != self.shape:
+            raise ValueError("All dimensions, except the first one,"
+                             " must match exactly")
+
+        self._offsets = np.r_[self._offsets, len(self._data)]
+        self._lengths = np.r_[self._lengths, len(element)]
+        self._data = np.append(self._data, element, axis=0)
+
+    def extend(self, elements):
+        """ Appends all `elements` to this compact list.
+
+        Parameters
+        ----------
+        elements : list of ndarrays, ``CompactList`` object
+            Elements to append. The shape must match already inserted elements
+            shape except for the first dimension.
+
+        """
+        if self._data.ndim == 0:
+            elem = np.asarray(elements[0])
+            self._data = np.zeros((0, elem.shape[1]), dtype=elem.dtype)
+
+        next_offset = self._data.shape[0]
+
+        if isinstance(elements, CompactList):
+            self._data.resize((self._data.shape[0]+sum(elements._lengths),
+                               self._data.shape[1]))
+
+            offsets = []
+            for offset, length in zip(elements._offsets, elements._lengths):
+                offsets.append(next_offset)
+                self._data[next_offset:next_offset+length] = elements._data[offset:offset+length]
+                next_offset += length
+
+            self._lengths = np.r_[self._lengths, elements._lengths]
+            self._offsets = np.r_[self._offsets, offsets]
+
+        else:
+            self._data = np.concatenate([self._data] + list(elements), axis=0)
+            lengths = list(map(len, elements))
+            self._lengths = np.r_[self._lengths, lengths]
+            self._offsets = np.r_[self._offsets, np.cumsum([next_offset] + lengths)[:-1]]
+
+    def copy(self):
+        """ Creates a copy of this ``CompactList`` object. """
+        # We do not simply deepcopy this object since we might have a chance
+        # to use less memory. For example, if the compact list being copied
+        # is the result of a slicing operation on a compact list.
+        clist = CompactList()
+        total_lengths = np.sum(self._lengths)
+        clist._data = np.empty((total_lengths,) + self._data.shape[1:],
+                               dtype=self._data.dtype)
+
+        next_offset = 0
+        offsets = []
+        for offset, length in zip(self._offsets, self._lengths):
+            offsets.append(next_offset)
+            clist._data[next_offset:next_offset+length] = self._data[offset:offset+length]
+            next_offset += length
+
+        clist._offsets = np.asarray(offsets)
+        clist._lengths = self._lengths.copy()
+
+        return clist
+
+    def __getitem__(self, idx):
+        """ Gets element(s) through indexing.
+
+        Parameters
+        ----------
+        idx : int, slice or list
```

Comment:
Yes. I'm not sure what to write though? It seems to me that if I put everything, the line would get too long. e.g.
`idx : int or slice or list of int or ndarray of bool dtype or ndarray of int dtype`